### PR TITLE
SI-6017 Fix issues around Scaladoc's alphabetical index page

### PR DESCRIPTION
**Pull request by kzys**
This pull request fix these two issues:
1. In the index page (index.html), the current implementation creates links in JavaScript but it generates dangling links.
2. In the alphabetical index page (index-$letter.html), the entries are not case-sensitive. For example, "STAR" and "Star" are listed under "STAR".

### CODE REVIEW

**Comment by heathermiller** [comment]
Hey Kato, thanks for the patch. This is actually something that I had been on the fence about how to solve.

I like that you resolved our issue with the index page by, in the left sidebar, not generating links for which there is no alphabetical index page. The issue, though, is that for small libraries with, say 10 classes or so, you end up with what looks like a random collection of a few letters in the left panel. If I were new to Scaladoc, it wouldn't be immediately obvious to me that that was an alphabetical index of entities/members.

The Java 7 API docs handle this a bit better, I think. They simply have a link (top of the right pane) called "Index" and at the top of the page, is the alphabetical index, and all entities/members for all available letters are listed. So there can be no confusion-- you know you're looking at an index instead of a strange collection of letters like "C F M S" with no idea that it's an index. Something like how Java does it might be clearest.

Also of note-- over the course of a few discussions with Vlad and some others on the doc team, we had come to the conclusion that we'd eventually like to do away with the index all together, and come up with a filter panel that also filters members. It was just chat, but we had talked about doing this for 2.11 at some point (and the current master branch is for 2.11)

This is all open to discussion of course.

Which release did you want to target this for? It could go into 2.10.1 as far as I can tell, in which case, it'd need to be pointed at the 2.10.x branch.

@VladUreche any thoughts on this?

**Comment by heathermiller** [comment]
Java 7 Api docs, for reference: http://docs.oracle.com/javase/7/docs/api/

**Comment by heathermiller** [comment]
Btw, those comments above pertain to the first of your two commits, c18984b. I'd just like to open the floor up to discussion about how to handle the index given that we plan on eventually getting rid of it anyway.

As far as the second commit goes, 50735ab, LGTM!

Thanks Kato!

**Comment by kzys** [comment]
I can agree with you. To be honest, I usually don't use the index to find members.

Also I implemented the "unified" filter to find both classes and methods in 2011. If you (and the doc team) haven't started working on that yet, I will take a look and implement a prototype.
http://www.scala-lang.org/node/8031

**Comment by VladUreche** [comment]
Hi everyone, thanks @kzys for the patch!

I think we should accept the current patch, and promise ourselves to work on removing the index altogether until 2.11. Right now we have dangling links, imho that's important to fix. Maybe we can strike out or shade the letters that are not part of the index instead of hiding them. WDYT?

**Comment by VladUreche** [comment]
By the way, this patch should target `2.10.x` not `master`.

**Comment by heathermiller** [comment]
Hey Vlad- could you have a look at the Java 7 docs with the "Index" link that I pointed to above? Did you see the concern that I had about some random letters not clearly being an index? As I said above, for small libraries I think it would be really confusing that it's even an index if it only shows the letters "C I M N S", for example.

To expand upon my earlier suggestion-- we could easily keep `def letters` and just put that at the top of the current index pages for each letter instead of generating the letter list in the left panel. Or, better, if it's not too much, we could combine all letters onto one page. Then, we just have to link to the "Index" in the left panel rather than displaying the whole alphabet there. I think it wouldn't be a lot of work, and it wouldn't throw away any of the current effort that kzys has already invested.

Regarding pointing it at 2.10.x, yes, I suggested that earlier.

So I think the best thing (for the sake of usability) would be to do one of the following:
1. do the above with a Java 7-like index page, or
2. since kzys has already implemented a prototype index search back in 2010 and is motivated to continue on that front, maybe we could just wait until that's ready? Then we don't even need the index at all.
3. otherwise, Vlad, your suggestion is a good one. Instead of striking-out or shading, we could just generate letters for the unused ones that aren't links. Just normal text.

**Comment by VladUreche** [comment]
> 1.  do the above with a Java 7-like index page, or
> 2.  since kzys has already implemented a prototype index search back in 2010 and is motivated to continue on that front, maybe we could just wait until that's ready? Then we don't even need the index at all.
> 3.  otherwise, Vlad, your suggestion is a good one. Instead of striking-out or shading, we could just generate letters for the unused ones that aren't links. Just normal text.

I'm hesitant to change the layout mid-release, so I would prefer to see 3 implemented in 2.10.1 and both of 1 and 2 for 2.11. WDYT?

**Comment by heathermiller** [comment]
Sure, I agree. We shouldn't change the layout of Scaladoc too much for 2.10.1, so we ought to only do 3 on the 2.10.x branch.

Though do you think that we should do both 1 and 2 for 2.11? If we come up with an intelligent way to do the filtering, we might be able to avoid 1 all together.

One idea: 
Attachment001

It's not the most beautiful sketch (sorry). But the idea is that there is simply nothing below the filter bar when no text is entered. When text is entered, three buttons below the filter bar appear, and in the results section, both members and entities which match the search string are shown. In this default case, the leftmost button "show all" is selected, and entities are shown first, then members are shown after. In the other two cases, we only either "show entities" or "show members". We can choose more intelligent button names.

But this seems like it could work without adding an additional page to have to click through.

@kzys, @VladUreche what do you think?

**Comment by VladUreche** [comment]
> It's not the most beautiful sketch (sorry). But the idea is that there is simply nothing below the filter bar when no text is entered. When text is entered, three buttons below the filter bar appear, and in the results section, both members and entities which match the search string are shown. In this default case, the leftmost button "show all" is selected, and entities are shown first, then members are shown after. In the other two cases, we only either "show entities" or "show members". We can choose more intelligent button names.

This looks very good. We need to make sure the JS search is fast, if not, we can switch back to "only entities" by default. Would you also keep the index page in this case?

**Comment by heathermiller** [comment]
Yep, that's what I meant to get at. If we were to do it this way, I don't think we'd need an index page. So we could do 2 and forget 1 all together.

**Comment by kzys** [comment]
Heather's suggestion sounds good. Then I will create an another pull request for 2.10.x and cherry-pick 50735ab.

**Comment by JamesIry** [comment]
What's the latest with this PR? 21 days with no activity. ping @kzys, @heathermiller, @VladUreche

**Comment by VladUreche** [comment]
Sorry, should have been closed because the fix in c18984b is not the optimal solution. 
@kzys, can you please reopen the pull request with all the letters showing in the index (but no links)?

**Comment by kzys** [comment]
@VladUreche Okay.

File: src/compiler/scala/tools/nsc/doc/model/IndexModelFactory.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@ object IndexModelFactory {
       object result extends mutable.HashMap[Char,SymbolMap] {
 
         /* symbol name ordering */
-        implicit def orderingMap = math.Ordering.String.on { x: String => x.toLowerCase }
```
Comment (dcsobral) [inline src/compiler/scala/tools/nsc/doc/model/IndexModelFactory.scala:21]: Personally, I'd have made it `{ x: String => (x.toLowerCase, x) }`, so they would _sort_ together but still be considered separate words, and have both words starting with lowercase and upper case listed together. I think it is more intuitive, and would list together things like `thisWord` and `ThisWord`, which is not uncommon.

**Pull request closed without merging**